src/database/database-internal.ts

```diff
@@ -96,9 +106,45 @@ export class DatabaseService {
 
       this.databases[dbUrl] = db;
     }
+
+    if (!this.tokenRefresh) {
+      this.tokenRefresh = true;
+      this.appInternal.INTERNAL.addAuthTokenListener(this.onTokenChange);
+    }
+
     return db;
   }
 
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  private onTokenChange(_: string): void {
+    this.appInternal.INTERNAL.getToken()
+      .then((token) => {
+        const delayMillis = token.expirationTime - TOKEN_REFRESH_THRESHOLD_MILLIS - Date.now();
+        // If the new token is set to expire soon (unlikely), do nothing. Somebody will eventually
+        // notice and refresh the token, at which point this callback will fire again.
+        if (delayMillis > 0) {
+          this.scheduleTokenRefresh(delayMillis);
+        }
+      })
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      .catch((_) => {
+        // Unlikely error, since a new token was just fetched before the callback fired.
+        // Just here to prevent the promise chain from crashing.
+      });
+  }
+
+  private scheduleTokenRefresh(delayMillis: number): void {
+    clearTimeout(this.tokenRefreshTimeout);
+    this.tokenRefreshTimeout = setTimeout(() => {
+      this.appInternal.INTERNAL.getToken(true)
```

Comment:
nit: I don't know if it conflicts with the style guide here, but style guides internally say you should comment to name any boolean variable names:

```js
this.appInternal.INTERNAL.getToken(/*forceRefresh=*/true);
```

Reply:
Done